dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java

```diff
@@ -173,65 +267,49 @@ public DoraLoadJob(String path, Optional<String> user, String jobId, OptionalLon
    */
   private List<DoraLoadTask> prepareNextTasks(Set<WorkerInfo> workers) {
     LOG.debug("Preparing next set of tasks for jobId:{}", mJobId);
+    mLoadSubTaskIterator.updateWorkerList(workers);
     int workerNum = workers.size();
     ImmutableList.Builder<LoadSubTask> batchBuilder = ImmutableList.builder();
-    if (mCurrentSubTaskIterator == null) {
-      if (mUfsStatusIterator.hasNext()) {
-        mCurrentSubTaskIterator = initSubTaskIterator(workers);
-      }
-      else {
-        return Collections.emptyList();
-      }
-    }
-    int i = 0;
-    int startRetryListSize = mRetrySubTasks.size();
-    int numSubTasks = 0;
-    while (numSubTasks < RETRY_THRESHOLD
-        && i++ < startRetryListSize && mRetrySubTasks.peek() != null) {
-      LoadSubTask subTask = mRetrySubTasks.poll();
-      String path = subTask.getUfsPath();
-      try {
-        mUfs.getStatus(path);
-        batchBuilder.add(subTask);
-        ++numSubTasks;
-      } catch (IOException | AlluxioRuntimeException e) {
-        // The previous list or get might contain stale file metadata.
-        // For example, if a file gets removed before the worker actually loads it,
-        // the load will fail and the scheduler will retry.
-        // In such case, a FileNotFoundException might be thrown when we attempt to
-        // get the file status again, and we simply ignore that file.
-        if (!(e instanceof FileNotFoundException || e instanceof NotFoundRuntimeException)) {
-          mRetrySubTasks.offer(subTask);
-        }
-      }
-    }
-    while (numSubTasks < BATCH_SIZE * workerNum) {
-      if (!mCurrentSubTaskIterator.hasNext()) {
-        if (!mUfsStatusIterator.hasNext()) {
-          break;
-        }
-        else {
-          mCurrentSubTaskIterator = initSubTaskIterator(workers);
+
+    for (int numSubTasks = 0; numSubTasks < BATCH_SIZE * workerNum; ++numSubTasks) {
+      if (mLoadSubTaskIterator.hasNext()) {
+        batchBuilder.add(mLoadSubTaskIterator.next());
+        numSubTasks++;
+      } else if (!mRetrySubTasksDLQ.isEmpty()) {
+        LoadSubTask subTaskToRetry = mRetrySubTasksDLQ.poll();
+        String path = subTaskToRetry.getUfsPath();
+        try {
+          if (GET_LATEST_UFS_ON_RETRY) {
```

Comment:
provide an option to avoid getting metadata again. getting an ufs status without any concurrency is low effcient. files to load (in AI use cases) barely change.